### PR TITLE
STM32 FDCAN: Get all messages from FIFO in RX ISR

### LINKED ISSUE #1013: Unittest failing fdcan_test on Nucleo-G474RE
The test: https://github.com/modm-io/modm/blob/389a9c352e9d20c1f8ee8a40d5276c2fd3df0557/test/modm/platform/fdcan/fdcan_test.cpp#L88-L100

<details>
<summary>
The output:

```
[...]
FAIL: fdcan_test:92 : 1801 == 18193␊
FAIL: fdcan_test:93 : 0 == 1␊
FAIL: fdcan_test:96 : 0 == 1␊
[...]
```

</summary>

```
>>> fdcan_test␊
FAIL: fdcan_test:62 : true == false␊
FAIL: fdcan_test:92 : 1801 == 18193␊
FAIL: fdcan_test:93 : 0 == 1␊
FAIL: fdcan_test:96 : 0 == 1␊
FAIL: fdcan_test:93 : 1 == 2␊
FAIL: fdcan_test:96 : 1 == 2␊
FAIL: fdcan_test:96 : 52 == 2␊
FAIL: fdcan_test:93 : 2 == 3␊
FAIL: fdcan_test:96 : 2 == 3␊
FAIL: fdcan_test:96 : 2 == 3␊
FAIL: fdcan_test:96 : 86 == 3␊
FAIL: fdcan_test:93 : 3 == 4␊
FAIL: fdcan_test:96 : 3 == 4␊
FAIL: fdcan_test:96 : 3 == 4␊
FAIL: fdcan_test:96 : 3 == 4␊
FAIL: fdcan_test:96 : 239 == 4␊
FAIL: fdcan_test:93 : 4 == 5␊
FAIL: fdcan_test:96 : 4 == 5␊
FAIL: fdcan_test:96 : 4 == 5␊
FAIL: fdcan_test:96 : 4 == 5␊
FAIL: fdcan_test:96 : 4 == 5␊
FAIL: fdcan_test:96 : 4 == 5␊
FAIL: fdcan_test:93 : 5 == 6␊
FAIL: fdcan_test:96 : 5 == 6␊
FAIL: fdcan_test:96 : 5 == 6␊
FAIL: fdcan_test:96 : 5 == 6␊
FAIL: fdcan_test:96 : 5 == 6␊
FAIL: fdcan_test:96 : 5 == 6␊
FAIL: fdcan_test:96 : 5 == 6␊
FAIL: fdcan_test:93 : 6 == 7␊
FAIL: fdcan_test:96 : 6 == 7␊
FAIL: fdcan_test:96 : 6 == 7␊
FAIL: fdcan_test:96 : 6 == 7␊
FAIL: fdcan_test:96 : 6 == 7␊
FAIL: fdcan_test:96 : 6 == 7␊
FAIL: fdcan_test:96 : 6 == 7␊
FAIL: fdcan_test:96 : 6 == 7␊
FAIL: fdcan_test:93 : 7 == 8␊
FAIL: fdcan_test:96 : 7 == 8␊
FAIL: fdcan_test:96 : 7 == 8␊
FAIL: fdcan_test:96 : 7 == 8␊
FAIL: fdcan_test:96 : 7 == 8␊
FAIL: fdcan_test:96 : 7 == 8␊
FAIL: fdcan_test:96 : 7 == 8␊
FAIL: fdcan_test:96 : 7 == 8␊
FAIL: fdcan_test:96 : 7 == 8␊
FAIL: fdcan_test:99 : true == false␊
FAIL: fdcan_test:100 : true == false␊
```

</details>

### PR DESCRIPTION
**Pull request by ser-plu**
1. In cases, when a new message comes during RX ISR execution, it would not be processed until a next message arrives. Critical for request-response protocols.
2. The header file missed a dependency

### CODE REVIEW

File: src/modm/platform/can/stm32-fdcan/can.cpp.in [hunk 1]
```diff
@@ -239,23 +239,23 @@ MODM_ISR({{ reg }}_IT1)
 %% if options["buffer.rx"] > 0
 	RxMessage rxMessage;
 
-	if (rxFifo0HasMessage()) {
+	while (rxFifo0HasMessage()) {
```
Comment (rleh) [inline src/modm/platform/can/stm32-fdcan/can.cpp.in]: A while loop bears the risk that the interrupt runs forever and as a result the user code never removes messages from `rxQueue`.
What do you think about something like the following?

<details>

<summary>Click to view code</summary>


```c++
MODM_ISR(FDCAN1_IT1)
{
	RxMessage rxMessage;

	int_fast16_t msgRetrieveLimit = rxQueue.getMaxSize() - rxQueue.getSize();

	while (rxFifo0HasMessage() && (msgRetrieveLimit > 0)) {
		readMsg(rxMessage.message, 0, &rxMessage.filter_id, &rxMessage.timestamp);
		modm_assert_continue_ignore(rxQueue.push(rxMessage), "fdcan.rx.buffer",
			"CAN receive software buffer overflowed!", 1);
		msgRetrieveLimit--;
	}
	// acknowledge interrupt flag
	FDCAN1->IR = FDCAN_IR_RF0N;

	while (rxFifo1HasMessage() && (msgRetrieveLimit > 0)) {
		readMsg(rxMessage.message, 1, &rxMessage.filter_id, &rxMessage.timestamp);
		modm_assert_continue_ignore(rxQueue.push(rxMessage), "fdcan.rx.buffer",
			"CAN receive software buffer overflowed!", 1);
		msgRetrieveLimit--;
	}
	// acknowledge interrupt flag
	FDCAN1->IR = FDCAN_IR_RF1N;
}
```


</details>

**Review by rleh — COMMENTED** [review]
Thanks for bringing this up!
I had noticed this code might be buggy when I wrote the [SAM MCAN driver](https://github.com/modm-io/modm/tree/bfafcd395103878a9dca3aff7fec31e83a895d55/src/modm/platform/can/sam-mcan) (based on this driver), but then forgot about it.

File: src/modm/platform/can/stm32-fdcan/can.cpp.in (hunk 1 above)
Comment (ser-plu) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: Then we also don't need the asserts anymore, right?
Otherwise seems as a nice solution

Comment (rleh) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: > Then we also don't need the asserts anymore, right?

We should still assert if the buffer is full at the beginning of the ISR, e.g.:
```c++
if (msgRetrieveLimit == 0) {
	modm_assert_continue_ignore(false, "fdcan.rx.buffer",
		"CAN receive software buffer full, not reading new message(s)!");
	return;
}
```

Comment (ser-plu) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: Adjusted the code.

File: src/modm/platform/can/stm32-fdcan/can.cpp.in [hunk 2]
```diff
@@ -237,25 +237,31 @@ MODM_ISR({{ reg }}_IT0)
...
 		readMsg(rxMessage.message, 1, &rxMessage.filter_id, &rxMessage.timestamp);
-		// acknowledge interrupt flag
-		{{ reg }}->IR = FDCAN_IR_RF1N;
-
-		modm_assert_continue_ignore(rxQueue.push(rxMessage), "fdcan.rx.buffer",
-			"CAN receive software buffer overflowed!", {{ id }});
+		rxQueue.push(rxMessage);
+		msgRetrieveLimit--;
 	}
+	{{ reg }}->IR = FDCAN_IR_RF1N;  // acknowledge interrupt flag
```
Comment (rleh) [inline src/modm/platform/can/stm32-fdcan/can.cpp.in]: There might still be messages in the FDCAN hardware FIFOs. Is it ok to acknowledge the interrupt flags?
(I think there is no better way...)

But acknowledging both flags with a single register write is faster and should have the same effect:

```c++

	{{ reg }}->IR = FDCAN_IR_RF0N | FDCAN_IR_RF1N;
```

Comment (ser-plu) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: If flush the FIFO, we just lose the messages. If we have them like now, then we still have a chance to read them at the next IRQ. So i think we need just to acknowledge here.

Comment (ser-plu) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: Maybe I can add a comment about this

Comment (rleh) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: Might be nice to be able to trigger the ISR at the end of the `getMessage()` function, but unfortunately I can't think of any way which would not require significant additional overhead due to mutexes.

**Review by rleh — APPROVED** [review]

File: src/modm/platform/can/stm32-fdcan/can.cpp.in (hunk 2 above)
Comment (ser-plu) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: I think, if the messages come that fast, that the rxQueue gets full, then the ISR will be triggered also quickly after the message is read :) Yes, the corner case of a big tight pack with a long silence aftewards is not covered, but that really requires a more complex logic

Comment (chris-durand) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: Couldn't we turn off the receive interrupt when the queue is full, not acknowledge the flag and turn the interrupt on again after the message has been retrieved? That way the interrupt would be retriggered automatically.

Comment (rleh) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: So `NVIC_DisableIRQ(FDCAN{{ id }}_IT1_IRQn)` in the `if (msgRetrieveLimit == 0)` clause instead of acknowledging the interrupt flag and `NVIC_EnableIRQ(FDCAN{{ id }}_IT1_IRQn)`  at the end of every `getMessage()` call?

**Comment by ser-plu** [comment]
Nice idea. I will add this tomorrow, as I have already left my workplace.

File: src/modm/platform/can/stm32-fdcan/can.cpp.in (hunk 2 above)
Comment (chris-durand) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: I was thinking of setting and clearing the `EINT1` bit in the `FDCAN_ILE` register. That could be implemented with a single store. `{{ reg }}->ILE = FDCAN_ILE_EINT1 | FDCAN_ILE_EINT0;` for turning on the interrupt and `{{ reg }}->ILE = FDCAN_ILE_EINT0;` for turning it off. `NVIC_EnableIRQ`/`NVIC_DisableIRQ` is probably slower because the implementation contains a branch and memory barrier intrinsics.

**Comment by rleh** [comment]
I implemented @chris-durand|s idea and also added a commit that fixes the fdcan unit test (see #1013).
The unittest still works...

**Comment by rleh** [comment]
>  The unittest still works...

... but not with the modification (7bede35) that actually triggers our corner case:

```
>>> LUDecomposition_test␊
>>> ad7280a_test␊
>>> angle_test␊
>>> arithmetic_traits_test␊
>>> atomic_queue_test␊
>>> bit_operation_test␊
>>> block_allocator_test␊
>>> bme280_test␊
>>> bmp085_test␊
>>> bounded_deque_test␊
>>> bounded_queue_test␊
>>> bounded_stack_test␊
>>> button_group_test␊
>>> button_test␊
>>> can_bit_timings_test␊
>>> can_connector_base_test␊
>>> can_connector_test␊
>>> can_lawicel_formatter_test␊
>><0x11>G<break x 002>
h<0x04><0x04>edddd<break x 005>
t<0xc7><0x16>>> circle_2d_test␊
>>> clock_test␊
>>> color_test␊
>>> cxa_guard_test␊
>>> debounce_test␊
>>> delay_test␊
<0x11>G<break x 002>
h<0x01><0x01>piddd<break x 005>
s,<0x17>>>> doubly_linked_list_test␊
>>> drv832x_<0x11>G<break x 002>
x<0x07><0x07>toooooooo<break>
m<0xd8>'_array_test␊
>>> endi<0x11>G<break x 002>
x<0x03><0x03>t{{{o<break x 005>
Gg>h<0x02><0x02>ezzz
```
WTF?

**Comment by ser-plu** [comment]
> I implemented @chris-durand|s idea and also added a commit that fixes the fdcan unit test (see #1013). The unittest still works...

I changed the code to use ILE register and adjusted the IRQ disable logic.

File: src/modm/platform/can/stm32-fdcan/can.cpp.in [hunk 3]
```diff
@@ -377,7 +375,7 @@ modm::platform::Fdcan{{ id }}::getMessage(can::Message& message, uint8_t *filter
 			(*timestamp) = rxMessage.timestamp;
 		}
 		rxQueue.pop();
-		NVIC_EnableIRQ({{ reg }}_IT1_IRQn);
+		{{ reg }}->ILE = FDCAN_ILE_EINT1 && FDCAN_ILE_EINT0; // reenable ISR, in case it was disabled due to rxQueue full
```
Comment (chris-durand) [inline src/modm/platform/can/stm32-fdcan/can.cpp.in]: You probably mean `FDCAN_ILE_EINT1 | FDCAN_ILE_EINT0` :wink: . `FDCAN_ILE_EINT1 && FDCAN_ILE_EINT0` evaluates to `true` which is converted to `1` on assignment.

Comment (ser-plu) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: 😳🙈🤦‍♂️🤦‍♂️🤦‍♂️
Wow. That's silly

Comment (ser-plu) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: Thanks

**Comment by chris-durand** [comment]
@rleh The modified `testBuffer()` unit test case writes 32 bytes into an 8 byte message data buffer:

```
for (uint_fast16_t i = 0; i <= numberOfMsgs; ++i) {
    message.setLength(i % 8);
    for (uint_fast8_t dataIndex = 0; dataIndex < i; ++dataIndex) {
        message.data[dataIndex] = i;
    }
    Fdcan1::sendMessage(message);
}
```
`dataIndex` goes up to `numberOfMsgs - 1` which is 31. No wonder that this is crashing.

File: src/modm/platform/can/stm32-fdcan/can.cpp.in (hunk 3 above)
Comment (ser-plu) [reply src/modm/platform/can/stm32-fdcan/can.cpp.in]: Fixed

**Comment by ser-plu** [comment]
I fixed the CAN tests as well

**Comment by rleh** [comment]
@salkinium What's wrong with the CI now?
It fails during the step _Test run of docs.modm.io-generator-script_ with `KeyboardInterrupt` exception in the _SpawnPoolWorker_:

<details>
<summary>Click to view log</summary>

```
[...]
Process SpawnPoolWorker-4:
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "/__w/modm/modm/tools/scripts/docs_modm_io_generator.py", line 219, in create_target
    symlinks[hash(kfile.read_bytes())].append(kfile)
  File "/usr/lib/python3.8/pathlib.py", line 1229, in read_bytes
    with self.open(mode='rb') as f:
  File "/usr/lib/python3.8/pathlib.py", line 1222, in open
    return io.open(self, mode, buffering, encoding, errors, newline,
KeyboardInterrupt
Traceback (most recent call last):
  File "tools/scripts/docs_modm_io_generator.py", line 143, in main
    results = pool.map(create_target, devices)
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 765, in get
    self.wait(timeout)
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 762, in wait
    self._event.wait(timeout)
  File "/usr/lib/python3.8/threading.py", line 558, in wait
    signaled = self._cond.wait(timeout)
  File "/usr/lib/python3.8/threading.py", line 302, in wait
    waiter.acquire()
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tools/scripts/docs_modm_io_generator.py", line 143, in main
    results = pool.map(create_target, devices)
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 736, in __exit__
    self.terminate()
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 654, in terminate
    self._terminate()
  File "/usr/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 692, in _terminate_pool
    cls._help_stuff_finish(inqueue, task_handler, len(pool))
  File "/usr/lib/python3.8/multiprocessing/pool.py", line 675, in _help_stuff_finish
    time.sleep(0)
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tools/scripts/docs_modm_io_generator.py", line 261, in <module>
    exit(main())
  File "tools/scripts/docs_modm_io_generator.py", line 161, in main
    return results.count(False)
  File "/usr/lib/python3.8/tempfile.py", line 966, in __exit__
    self.cleanup()
  File "/usr/lib/python3.8/tempfile.py", line 970, in cleanup
    self._rmtree(self.name)
  File "/usr/lib/python3.8/tempfile.py", line 952, in _rmtree
    _rmtree(name, onerror=onerror)
  File "/usr/lib/python3.8/shutil.py", line 718, in rmtree
    _rmtree_safe_fd(fd, path, onerror)
  File "/usr/lib/python3.8/shutil.py", line 655, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  File "/usr/lib/python3.8/shutil.py", line 655, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  File "/usr/lib/python3.8/shutil.py", line 655, in _rmtree_safe_fd
    _rmtree_safe_fd(dirfd, fullname, onerror)
  [Previous line repeated 3 more times]
  File "/usr/lib/python3.8/shutil.py", line 673, in _rmtree_safe_fd
    os.unlink(entry.name, dir_fd=topfd)
KeyboardInterrupt
```

</details>

**Comment by salkinium** [comment]
> with `KeyboardInterrupt` exception

Wow, impressive. Maybe the NSA guy assigned to us got annoyed by all the dots, who knows…

Multiprocessing in Python is just utterly broken. It's an underlying issue of fork() vs. globals initialization, that cannot really be fixed. You can replace it with Threadpool + subprocess.run, which effectively does the same, but with a controlled entry point back into the script and thus properly initialized globals.

**Comment by chris-durand** [comment]
Running the hardware unit test still fails. I'm currently investigating the issue.

**Comment by chris-durand** [comment]
I have found some issues with the TX handling. `sendMsg()` was checking for the software queue to be not full before sending instead of the hardware buffers. Also there was a race condition. `sendMsg()` could be called concurrently from `sendMessage()` and the TX interrupt handler. Both could write to the same TX buffer. The solution was to disable CAN interrupts in `sendMessage()`. Now the hardware test succeeds.

**Comment by salkinium** [comment]
(There's still a race-condition in the docs generator deduplication code, I'll look at it on Sunday)

**Comment by salkinium** [comment]
Ok, I've fixed the docs generator, the PR now needs to be rebased and squashed.

**Review by chris-durand — APPROVED** [review]

**Review by salkinium — APPROVED** [review]
